### PR TITLE
Add utf-8 to reason not liked;Avoid divide by zero;Add missing functions: Capture all followed and unfollowed ever

### PR DESCRIPTION
**Pull request by sionking**
Good when we handle real users and facing any bug so we can't unfollow anymore since the pool is empty.

### CODE REVIEW

File: instapy/print_log_writer.py [hunk 1]
```diff
@@ -24,3 +24,35 @@ def log_followed_pool(login, followed, logger, logfolder):
...
+    # We save all followed to a pool that will never be erase
+    log_record_all_followed(login, followed, logger, logfolder)
+
+
+def log_uncertain_unfollowed_pool(login, unfollowed, logger, logfolder):
+    """Prints and logs the uncertain unfollowed to
+    a seperate file"""
+    try:
+        with open('{0}{1}_uncertain_unfollowedPool.csv'.format(logfolder, login), 'a+') as followPool:
+            followPool.write(unfollowed + ",\n")
```
Comment (timgrossmann) [inline instapy/print_log_writer.py]: Could we use .format() here aswell, please?

File: instapy/print_log_writer.py [hunk 2]
```diff
@@ -24,3 +24,35 @@ def log_followed_pool(login, followed, logger, logfolder):
...
+    except BaseException as e:
+        logger.error("log_followed_pool error {}".format(str(e)))
+
+
+def log_record_all_unfollowed(login, unfollowed, logger, logfolder):
+    """Prints and logs the uncertain unfollowed to
+    a seperate file"""
+    try:
+        with open('{0}{1}_record_all_unfollowed.csv'.format(logfolder, login), 'a+') as followPool:
+            followPool.write(unfollowed + ",\n")
```
Comment (timgrossmann) [inline instapy/print_log_writer.py]: Could we use .format() here aswell, please?

File: instapy/print_log_writer.py [hunk 3]
```diff
@@ -24,3 +24,35 @@ def log_followed_pool(login, followed, logger, logfolder):
...
+    except BaseException as e:
+        logger.error("log_followed_pool error {}".format(str(e)))
+
+
+def log_record_all_followed(login, followed, logger, logfolder):
+    """We save all followed to a pool that will never be erase"""
+    try:
+        with open('{0}{1}_record_all_followed.csv'.format(logfolder, login), 'a+') as followPool:
+            followPool.write(followed + ",\n")
```
Comment (timgrossmann) [inline instapy/print_log_writer.py]: Could we use .format() here aswell, please?

**Review by timgrossmann — APPROVED** [review]

File: instapy/print_log_writer.py (hunk 3 above)
Comment (sionking) [reply instapy/print_log_writer.py]: there is a format().. I dont get it...

**Comment by CharlesCCC** [comment]
I'm getting the following error with the change you implemented (avoid divide by zero) : 

```
Traceback (most recent call last):
  File "quickstart-bypass.py", line 80, in <module>
    main(sys.argv[1:])
  File "quickstart-bypass.py", line 73, in main
    session.like_by_tags(amount=int(likeamount), use_smart_hashtags=True)
  File "C:\Users\Charles\OneDrive\InstaPyProd\instapy\instapy.py", line 678, in like_by_tags
    for i, link in enumerate(links):
TypeError: 'bool' object is not iterable
```

**Comment by sionking** [comment]
@CharlesCCC 
Hi I don't get how you get from like_by_tags  to get_links_for_username
And we always check:
            if links is False:
                continue
So please tell me what you see...

**Comment by CharlesCCC** [comment]
@sionking I'm not sure buddy, I first changed my local file to the one you made in `like_util.py`
and I'm getting the error I posted, then later on I have to change the code like the following, it seems worked so far:(temp. solution)
```
    link_elems = main_elem.find_elements_by_tag_name('a')
    total_links = len(link_elems)
    if total_links == 0:
        total_links = 1;
    links = []
```

**Comment by sionking** [comment]
I am not sure I hate this github merging please check you merged right and not just switched files.

There are only 2 places with this call and both check for 'False':
```
            try:
                links = get_links_for_username(
                    self.browser,
                    username,
                    amount,
                    self.logger,
                    randomize,
                    media)
            except NoSuchElementException:
                self.logger.error('Element not found, skipping this username')
                continue

            if (self.do_follow and
                username not in self.dont_include and
                following and
                    self.follow_restrict.get(username, 0) < self.follow_times):
                followed += follow_user(self.browser,
                                        self.follow_restrict,
                                        self.username,
                                        username,
                                        self.blacklist,
                                        self.logger,
                                        self.logfolder)
            else:
                self.logger.info('--> Not following')
                sleep(1)

            if links is False:
                continue
```

```
            try:
                links = get_links_for_username(self.browser,
                                               username,
                                               amount,
                                               self.logger,
                                               randomize,
                                               media)
            except NoSuchElementException:
                self.logger.error('Element not found, skipping this username')
                continue

            if links is False:
                continue
```

...
Can someone try?

**Comment by sionking** [comment]
Now I get it I have this check but latest version don't I will PR

**Comment by uluQulu** [comment]
In `like_image()` function of `like_util.py`
lines from 561 to 569, after latest commit of @sionking,
should be replaced with
```python
        elif len(liked_elem) == 1:
            logger.info('--> Image already liked! ')
            return False
        
        else:
            logger.info('--> Invalid Like Element!')
            return False
```
i noticed while merging, that, without this 👆 it will not notify correctly once encountered an **invalid like element**...

**Comment by sionking** [comment]
@uluQulu hi
```
def like_image(browser, username, blacklist, logger, logfolder):
    """Likes the browser opened image"""
    like_elem = browser.find_elements_by_xpath(
        "//a[@role='button']/span[text()='Like']/..")
    if len(like_elem) == 1:
        # sleep real quick right before clicking the element
        sleep(2)
        click_element(browser, like_elem[0])
        # check now we have unlike instead of like
        liked_elem = browser.find_elements_by_xpath(
            "//a[@role='button']/span[text()='Unlike']")
        if len(liked_elem) == 1:
            logger.info('--> Image Liked!')
            update_activity('likes')
            if blacklist['enabled'] is True:
                action = 'liked'
                add_user_to_blacklist(
                    browser, username, blacklist['campaign'], action, logger, logfolder
                )
            sleep(2)
            return True
        else:
            # if like not seceded wait for 2 min
            logger.info('--> Image was not able to get Liked! maybe blocked ?')
            sleep(120)
    else:
        liked_elem = browser.find_elements_by_xpath(
            "//a[@role='button']/span[text()='Unlike']")
        if len(liked_elem) == 1:
            logger.info('--> Image already liked! ')
            return False

    logger.info('--> Invalid Like Element!')
    return False
```

When invalid it returns False, So I don't see your point ...

**Comment by uluQulu** [comment]
@sionking 
Yea you're right, checked again, that's a crazy design xD

**Comment by sionking** [comment]
Hi, so is it working for you guys ?

**Comment by vladislav-vasilenko** [comment]
from .print_log_writer import log_uncertain_unfollowed_pool
ImportError: cannot import name 'log_uncertain_unfollowed_pool'

How to solve this?

**Comment by sionking** [comment]
@applehawk go and merge this PR
OR 
comment few lines

**Comment by timgrossmann** [comment]
@sionking Could you please change the mentioned lines to use `.format(...)`?

**Comment by sionking** [comment]
@timgrossmann Sorry I don't get it... It all uses .format()

**Comment by timgrossmann** [comment]
@sionking I commented the lines in your code:

```bash
with open('{0}{1}_uncertain_unfollowedPool.csv'.format(logfolder, login), 'a+') as followPool:
            followPool.write(unfollowed + ",\n")
```

**Comment by sionking** [comment]
@timgrossmann
I see the line you mentined but I don't understand what you want me to change ?
I use the .format in all lines.

**Comment by timgrossmann** [comment]
@sionking No you don't 

you use
```
followPool.write(unfollowed + ",\n")
```

to write to the file instead of something like

```
followPool.write('{},\n'.format(unfollowed))
```

**Comment by sionking** [comment]
Hi, 
@timgrossmann I now understand I will change it now.
But JFYTK it was like that in function log_followed_pool:
```
def log_followed_pool(login, followed, logger, logfolder):
    """Prints and logs the followed to
    a seperate file"""
    try:
        with open('{0}{1}_followedPool.csv'.format(logfolder, login), 'a+') as followPool:
            followPool.write(followed + ",\n")
    except BaseException as e:
        logger.error("log_followed_pool error {}".format(str(e)))
```

So I left it in the same format

**Comment by sionking** [comment]
Done

**Comment by timgrossmann** [comment]
@sionking Awesome!

Could we please get 2 or at least 1 more approval here?

**Comment by uluQulu** [comment]
encoding `reason` to UTF-8 will solve possible exceptions once inappropriate word contains a non-unicode char (e.g. contains `ö`), I did this by just encoding `inapp_unit` in `check_link` but this is a more general solution for future releases.
Everything works **fine**, but I wonder one thing, @sionking 
what are those newly written three files for?
`_record_allfollowed`, `_record_allunfollowed`, `_uncertain_unfollowedPool`

do we **really** need those?

**Comment by sionking** [comment]
@uluQulu hi
look at:
https://github.com/timgrossmann/InstaPy/issues/1420

This just saved me.
I could have take the list from blackList or logs but it is easier that way.
We can add a flag for it...

Why more files is an issue ? size?

**Comment by uluQulu** [comment]
Right, stability is much important
for me, I  just started to use `multi_logs=True` parameter to have a clean **logs** folder :-)

**Comment by CharlesCCC** [comment]
@uluQulu agree with you. 
We should keep the flow simple and clean to achieve whatever function we would like to achieve. "more not always be good"   👍

**Review by uluQulu — APPROVED** [review]

**Comment by sionking** [comment]
After 3 different real accounts found with incorrect following pool I found the only solution is to log all activity, so I can manually clean those account.

If you prepare I can remove all logs but debug will be much harder/ or we can save all pools to a pool folder ....

**Comment by sionking** [comment]
@timgrossmann can you decide ?

File: instapy/print_log_writer.py (hunk 3 above)
Comment (estebancortero) [reply instapy/print_log_writer.py]: I suppose, he refers to inside the `write`

**Comment by timgrossmann** [comment]
@sionking Let's merge it for now... I hope at one point we can start a big refactor that makes things more clean.
We then should keep discussions like this open for longer.

What I'm wondering about at the moment however is if that is correct:

```python
with open('{0}{1}_record_all_followed.csv'.format(logfolder, login), 'a+') as followPool:
            followPool.write('{},\n'.format(followed))
```

Do you actually want the name of the csv file to be "<logfolder><login>"?
shouldn't the log folder be the path to a log folder?

Like this?
```python
with open('{0}/{1}_record_all_followed.csv'.format(logfolder, login), 'a+') as followPool:
            followPool.write('{},\n'.format(followed))
```

**Comment by uluQulu** [comment]
self.logfolder handles it while `multi_logs==True`
```python
self.logfolder = './logs/'
if multi_logs is True:
    self.logfolder = './logs/{}/'.format(self.username)
```
I think it is better for user once logging is dealt by `multi_logs` arg to provide flexibility

**Comment by sionking** [comment]
Sorry I don't use multi_logs, is mine correct ?

**Comment by sionking** [comment]
BTW, the change of like image will save alot of time since looking for xpath that not exist is a 20 sec issue per image ! so Now it should be much faster for you all.

**Comment by uluQulu** [comment]
if you don't use multi_logs your `logfolder` will be `self.logfolder = './logs/'`, so that every time you start a new username from the same instapy folder, will save logs into that `./logs/` folder instead of `./logs/myusername1/`
it is fine to not enable `multi_logs` but vice verse it is much more cleaner...

**Comment by sionking** [comment]
@uluQulu I don't run on a server I have real account running from each one own laptop/PC so it uses home IP address (safer) plus no expenses for server/proxy for me.
All files are sync using file sharing cloud.

**Comment by timgrossmann** [comment]
@sionking Alright, looks like everything is right then.
Let's merge this and make people happy!

Thank you @sionking for this contribution.

**Comment by uluQulu** [comment]
@sionking 
i see, i also run locally except multiple accaunts from same instapy folder, that's why having multiple log folders worths using

**Comment by jeremycjang** [comment]
Pulled these commits and it seems to have broken the unfollow program. I created an issue for it #1436 and will be reverting to the previous commit without these additional lists in the meantime.

**Comment by timgrossmann** [comment]
@sionking just found a bug when using upper follower count:

```
--> Image not liked: b'None'
```

Edit: looks like it's not only when using upper follower count but a lot of others aswell

**Comment by uluQulu** [comment]
@timgrossmann 
if im not mistaken, you're pointing out `b''` bytestring notation which appears in Python 3 consoles
it is very annoying for me, and i thing it is annoying for those who do data analysis from logs, too, cos logs also include that stamp **b'foobar'**
it's happening after encoding the raw output string to utf-8 bytes, since we can't quit utf-8 for reliability, we can remove it's `b''` stamp from output just by slicing encoded bytestring before sending the message to logger
i've posted a solution to this in #1413 (works for both Python 2 & 3)

**Pull request merged**